### ISSUE TITLE
Removing habitrpg-shared?

### ISSUE THREAD

**Issue by lefnire** [HabitRPG/habitica#3439]
Hey fellow devs! Mobile and web use the habitrpg-shared repo (art, css, algos, directives, services, etc). 

Well I've been overhauling mobile big time. Turns out I can also use many of the website's Jade files in mobile. To do this, I set a jade variable `mobile=true`, and the jade will compile alternate html/classes depending on if we're mobile or web. Saves a lotta time! My first thought was "let's move all these jade files to habitrpg-shared". However, I'm concerned that habitrpg-shared will end up housing just about everything, and make tying related commits b/w web & shared a nightmare, and cause a confusing directory structure.

So here's my thought. Let's kill habitrpg-shared, move everything back into the website. The mobile app will install the **website** as an npm dependency, and will use `gulp` (like `grunt`) to copy all necessary files into it's build. This will make developing the website much simpler, and the website is the primary property anyway. However, if any other properties are using habitrpg-shared, this could cause issues or unnecessary bloat. Some 3rd-partyists who may be using habitrpg-shared: @Alys @goldbattle @magicmicky, would this present a problem?

So pros/cons:
- Pro: Easier dev of website & mobile, less commit-pairing.
- Pro: Better organization
- Con: Bloat for any other apps using habitrpg-shared

@deilann @snicker @colegleason @SabreCat @paglias eta al

**Comment by snicker** [comment]
Completely agree with this direction.

**Comment by thepeopleseason** [comment]
:+1:

**Comment by deilann** [comment]
So many :thumbsup:

**Comment by goldbattle** [comment]
I don't think that the chrome extension should be affected. Just pulling from the main api I thought. The big important thing is the possibility of getting a api call for a character profile image, if we are not able to get the image assets as easily. Not sure if that is an issue, but it would still be nice to have a link in the api to a compiled image of user.

Good idea, sounds good! :+1:

**Comment by markusholdings** [comment]
Agreed!  The test functions will be easier to manage that way as well.

**Comment by Alys** [comment]
I use only the API, not habitrpg-shared. What you describe, lefnire, sounds very sensible.

**Comment by MagicMicky** [comment]
No problem for me at all, since I only use the API! :-)

**Comment by paglias** [comment]
:+1:

**Comment by paglias** [comment]
@lefnire I'm all for this change but what about coffeescript? Keeping js and coffee in the same repo would be a bit strange

**Comment by lefnire** [comment]
@paglias very true... should we convert it to JS? Or just keep everything as is in a subdirectory, just so they're on the same repo?

**Comment by paglias** [comment]
@lefnire I would go for an all js repo... content.coffee wouldn't change much, index.coffee more but then its code would be similar to src/controllers so i think it's fine.

Do you already have an idea on folder structure?

**Comment by lefnire** [comment]
@paglias haven't really thought too much on it (this ticket isn't high prio, unless you think otherwise). Only think is I'k the stuff that's obviously meant to be common to multiple projects (aka, what's already in habitrpg-shared) to be silo'd as much as possible.

**Comment by SabreCat** [comment]
Not having to deal with habitrpg-shared would make quite a bit of dev work much easier!

I would be sad to see CoffeeScript go away, though. Being "a bit strange" doesn't sound like sufficient reason to change course there.

**Comment by paglias** [comment]
A bit strange means: many functions in habitrpg-shared are core parts of the api and of the server with the only difference that they are executed on the client too. Having to change language when editing files in the same scope doesn't look the right thing to do

**Comment by thepeopleseason** [comment]
Having dealt with many a repo where there's a variety of languages, I don't think there's anything wrong with having coffee and js in the same repo, as long as they're wisely organized.

**Comment by deilann** [comment]
I have learned to love CS in many ways.

I think it also helps users who know how to code, but aren't 100% on their
JS.

**Comment by Gaelan** [comment]
I think we should be moving more towards coffee, personally.

**Comment by lefnire** [comment]
Ughh.. JS vs CS... amirite @litenull @paglias @SabreCat amirite amirite?

**Comment by paglias** [comment]
I personally like coffee script syntax but I have actually moved away from it and not for the additional build step it requires. I find it a small barrier to entry for new developers and personally I hate the fact that it's trying to be backward compatible up to IE6 (ES3) preventing some useful features like the get/set syntax for getters/setters for objects to be implemented. And it will only get worse as features from ES6 are slowly being implemented in browsers and especially in node where a lot of the new syntax can be already used with node 0.11 (give a look at co and koa for some really cool example). When debugging without additional plugin in node I don't like seeing stack traces' lines from the compiled js instead of the original coffee code.

Then, I understand that in algos where it's mostly math coffee can be a little useful but moving the whole project to coffee...

**Comment by paglias** [comment]
Keeping the status quo, it's probably ok, James is right: js remains js and cs from shared remains cs

**Comment by lefnire** [comment]
Right there with Matteo. Habit's had at least 10 big CS v js convos
throughout the years. The reason for moving from CS to js is primarily
contributors. All js devs know js, not all know CS. Some contribs almost
jumped in, blocked since they don't know CS. Others can figure it out, but
blocked by all the tooling (spotty debugging, error line numbers, etc.) I
think we can all agree that CS > js. Saves time, lines, and heartache.
Reads better. Protects the writer from js common pitfalls. But alas, it
doesn't bode as well with open source projects - better save it for
personal or internal projects. Regarding habitrpg-shared, I don't think we
should do anything at this point - unnecessary effort to convert it, let's
just drop it into a folder of the website and convert it if we find real
need to later.

**Comment by lefnire** [comment]
Thinking on the dir structure, my vote is:

```
habitrpg
--common
--mobile
--website
--chrome
```

I spoke with someone at a meetup who has a similar setup as ours, similarly moved from 3 repos to 1. He said "make sure to put mobile in there, or you'll face the same issues" (commit-pairing, tests-pairing, etc). 

`require('habitrpg-shared')` won't have any issues as `require('../../../common')`. The only problem I see is `bower_components/habitrpg-shared`. Probably the best bet is a copy dir in the build process of both web & mobile. 

Or should we keep the top level dirs as is, move common into a nested dir of web, and make mobile a subdir of website? (aka, keep repo as is now, drop mobile/common into repo)

**Comment by lefnire** [comment]
Oh, and I just realized... we're gonna want to merge most (if not all) PRs on mobile & shared repos. And if we do No. 1 dir structure above, same goes for web. And what to do about pending issues.... oi, let's postpone this puppy

**Comment by Ulugbeg** [comment]
Hi! As a potential contributor who does NOT know CoffeeScript, I was looking for more info on the plan cited on [Contributing to HabitRPG](http://habitrpg.wikia.com/wiki/Contributing_to_HabitRPG#Side-note_on_technology):

> We're moving from CoffeeScript to JavaScript because CS is a bigger barrier-to-entry for Habit as an open-source project.

Hear Hear! : ) 

Now, this particular issue seems as close as a plan as I could find in that direction, but there does not seem to be a concrete plan for it - so I assume there is no schedule as to when this might be happening? So if I do want to contribute, I actually have to learn CoffeeScript for the (near?) future?

Oh, and if there is an issue for this, I apologize and would love a pointer to it :)

**Comment by Gaelan** [comment]
Are we really moving to JS? :disappointed:

**Comment by lefnire** [comment]
> We're moving from CoffeeScript to JavaScript because CS is a bigger barrier-to-entry for Habit as an open-source project.

Dunno who added that in, but it was without asking any of the devs. We've had this convo _very very_ extensively and every time we have it, CS wins.

**Comment by Gaelan** [comment]
Yay!

EDIT: Just fixed the wiki.

**Comment by lefnire** [comment]
@Ulugbeg that might be old regarding the website. Used to be _everything_ was in CS. We moved the website to JS, since that's where most people were contributing. habitrpg-shared, however, is quite heavily algorithmic. If you look at the type of code there, you'll see why it benefits from CS. Same with the tests, we merged a PR which moved shared's tests to JS - thinking "this will make life much harder, for a theoretical onboarding improvement" and that's exactly what it did. 

Hope this doesn't sound like the smack-down, it's still an open question - but each time we have the debate CS wins, so we're looking for better reasons than on-boarding at present

**Comment by Ulugbeg** [comment]
Ok, well that's cleared up then :) Time for me to learn a new language I guess...

**Comment by lefnire** [comment]
I'll be looking into this soon. We'll want to merge as many habitrpg-mobile & habitrpg-shared PRs as possible, close those that don't apply, etc. For the very extensive ones, the diffs/patches will still be available to work with via https://developer.github.com/v3/pulls/ (eg, https://github.com/HabitRPG/habitrpg-shared/pull/373.patch) to use against the main repo as necessary. Eventually, let's also move the chrome extension to the single repo - but not super prio

**Comment by paglias** [comment]
@lefnire do you want to keep a main directory for each of the repos and leave the rest unchanged?

**Comment by lefnire** [comment]
I'm thinking along the lines of https://github.com/HabitRPG/habitrpg/issues/3439#issuecomment-43124378 , is that what you're asking about?

**Comment by paglias** [comment]
@lefnire exactly :)

**Comment by taldin** [comment]
So essentially, if it was habitrpg-shared, it becomes 'common'?     And will this be in the 'node_modules' directory or somewhere else?     Will you still need to run grunt on it to update changes?

**Comment by lefnire** [comment]
it _won't_ be in node_modules. The module will be accessed as `require('../../common')` (or however many `..`s are required at that location). The only hiccup will be accessing it in the public (client) directory. We don't want a symlink, there are couple solutions:
1. Web: [connect-assets](https://github.com/adunkman/connect-assets) can pull external directories into public cat/min js files. connect-assets also has smart caching, watching, and other features (cc @paglias), which now may be a good time to migrate a lot of `grunt`'s work to this module
2. Mobile: We'll just need to copy the necessary files in `gulp`.

In the first pass, you'll still need to run `grunt` & `grunt full` as necessary to update common's files. Obviously that's not ideal, but adding a pre-start script to run those for web or mobile is whole other task, so we'll keep them separate for now.

**Comment by taldin** [comment]
Hm.  I'm not sure I understand.  If you say it 'won't be in node_modules',  but then you go on to say that it will be accessed as 'require(../../node_modules)'    even though it isn't in there?  Or is it just that habitrpg-shared will live in common, but other things in node_modules are still needed?

**Comment by crookedneighbor** [comment]
@taldin The reason that habitrpg-shared exists is because it was thought that it was necessary in order to have shared files between the main repo and the mobile repo. We figured out a way to be able to share things to the mobile repo just using the main repo, so it's no longer necessary.

In node, if you write `require('module-name')`, it'll look for the script in the node-modules folder, but if you write `require('../path/to/script')`, it'll load it from that location instead.

Grunt will still be used to compile the coffee script and do the css preprocessing, but likely the new common script will be added to the grunt task that gets kicked off on npm start.

Does that make sense?

**Comment by Alys** [comment]
@taldin 'require(../../node_modules)' was just a typo. It will be 'require(../../common)'

**Comment by lefnire** [comment]
what they said. Yeah, that typo threw you off - sorry about that - `require('../../common')` - important distinction

**Comment by taldin** [comment]
Gotcha, reading back makes more sense now. :)

**Comment by crookedneighbor** [comment]
@lefnire Is there anything specific you want me to do for this? If the answer is "Close as many issues on the shared and mobile repos as possible" that's fine.

**Comment by lefnire** [comment]
as always, thanks a million @crookedneighbor! yeah, merging or closing PRs on habitrpg-shared is the first step. If it's something that's too big yet unfinished, get a link to the PR's patch, create a new issue on habitrpg with that link, or whatever steps you think would be best

**Comment by crookedneighbor** [comment]
@lefnire and @paglias Just finished migrating all the open issues from shared to the main repo (will work on PRs later).

Should the [issues feature be disabled](https://github.com/HabitRPG/habitrpg-shared/settings) now for the shared repo?

**Comment by paglias** [comment]
@crookedneighbor good idea, done and added a link to the main repo in the readme

**Comment by crookedneighbor** [comment]
Down to 6 PRs:

[My Mount Master/Triad Bingo Achievement](https://github.com/HabitRPG/habitrpg-shared/pull/388) (ready to be merged, just needs a Bailey announcement)

3 PRs from @Alys 
[allow Quest Scrolls to be sold for Gold](https://github.com/HabitRPG/habitrpg-shared/pull/417)
[create a "Dated" To-Dos tab](https://github.com/HabitRPG/habitrpg-shared/pull/350)
[change behaviour of chain quests (scrolls to owner only)](https://github.com/HabitRPG/habitrpg-shared/pull/303)

@negue's [newest PR for markdown timeoutTime attribute](https://github.com/HabitRPG/habitrpg-shared/pull/441), which I assume will be merged soon.

@lefnire's [WIP Undo Feature](https://github.com/HabitRPG/habitrpg-shared/pull/95)

**Comment by Alys** [comment]
Ah geeeeez....

**Comment by Alys** [comment]
[change behaviour of chain quests (scrolls to owner only)](https://github.com/HabitRPG/habitrpg-shared/pull/303)  --- This one will need a Bailey announcement too, when it's ready, which will be soon. Any day now.

**Comment by crookedneighbor** [comment]
Correct me if I'm wrong, but I believe @paglias suggested moving the public folder into common. I think that's a good idea, so I'm doing that.

If anyone objects, let me know.

**Comment by paglias** [comment]
@crookedneighbor I would keep site specific files in this repo, manily public/css and public/js and maybe public/fontello but I'm not sure

**Comment by crookedneighbor** [comment]
# Progress report:

### Overview/Checklist of lingering things/TL;DNR

I've set up website specific files to live in the website directory and shared files to live in the common directory. 

~~I have some files (audio, sprites, js scripts) symlinked from common -> website/public, is that good practice? Or should it copy them on server start?~~
Edit: Obviously it wasn't a good practice since the symlinks weren't part of the repo. :anguished: 

**Still needs doing:**
- [x] Add shared Grunt tasks
- [x] Correct paths in mobile repo to use files in website and common
- [x] Merge/Close new prs on shared repo
- [x] Make sure vagrant command still works
- [x] Update README.md and other md files

### Gulp

I've gotten almost all the tasks converted from grunt -> gulp. I'm having trouble with the hashres task, mostly because there isn't a version of the module made by the same group for gulp, so not sure which of the gulp plugins that do the same thing will meet our needs. @paglias if you have input, I'd love to hear it.

[Here's the current gulp file](https://github.com/HabitRPG/habitrpg/blob/common-convert/gulpfile.js).

### Mobile

Haven't done any work to correct paths for mobile to use the common or website directories where applicable.

The views folder is kind of tricky, since there are views that include mixins that mobile uses. Right now it lives in the website directory. Should I go through and move any file that contains a mixin with mobile as an argument to common? Move all of the views to common? Or leave all the views in website?

### Vagrant

Haven't checked to see if anything needs to be changed on the Vagrant file to be updated for this new organization.

### Directory structure (two levels deep, except website, common, and test)

Anything that could be organized better here? Can we remove some files?

``` bash
├── CHANGELOG.md
├── CONTRIBUTING.md
├── DOCS-README.md
├── Dockerfile
├── EXTENDEDCHANGELOG.md
├── Gruntfile.js # Still present for reference while converting to gulp
├── LICENSE
├── Procfile
├── README.md
├── VAGRANT.md
├── Vagrantfile
├── archive
│   ├── README.md
│   └── derby_controllers
├── bower.json
├── common
│   ├── README.md
│   ├── audio
│   │   ├── danielTheBard
│   │   └── wattsTheme
│   ├── css
│   │   ├── backer.css
│   │   ├── css.template.mustache
│   │   └── index.css
│   ├── dist # Ignored in .gitignore, build process for prod may want to run browserify and script task before starting the server
│   │   ├── scripts # just the browserified habitrpg-shared.js file right now
│   │   └── sprites
│   ├── img
│   │   ├── bs-docs-masthead-pattern.png
│   │   ├── calendar_google.png
│   │   ├── calendar_ical.png
│   │   ├── emoji
│   │   ├── logo
│   │   ├── project_files
│   │   ├── screenshot.jpeg
│   │   └── sprites
│   ├── index.js
│   ├── locales
│   │   └── (language json files) # full list omitted 
│   └── script
│       ├── config.js
│       ├── content.coffee
│       ├── directives.js
│       ├── i18n.coffee
│       ├── index.coffee
│       └── userServices.js
├── database_reports
│   ├── count_users_who_own_specified_gear.js
│   └── duplicate_indexes.js
├── gulpfile.js
├── karma.conf.js
├── keys
│   └── google
├── migrations
│   └── (Migration Scripts) # full list of scripts omitted
├── newrelic.js
├── newrelic_agent.log
├── node_modules
│   └── (Node Modules) # full list of modules omitted
├── package.json 
├── protractor.conf.js
├── register-handlers.js # Config file for coverage from shared. Only covers scripts in common/script, so could be moved into common directory
├── test
│   ├── README.md
│   ├── api.mocha.coffee
│   ├── archive
│   ├── common # habitrpg-shared tests moved to this directory
│   │   ├── algos.mocha.coffee
│   │   ├── mocha.opts
│   │   ├── simulations
│   │   └── test_helper.coffee
│   ├── e2e
│   ├── mocha.opts
│   ├── run_tests.sh
│   └── spec
├── vagrant.sh
├── validate-commit-msg.js
├── validate-commit-msg.spec.js
└── website
    ├── build
    ├── config.json # website/src/server.js was looking for config.json in the website directory, so I moved it here
    ├── config.json.example
    ├── public
    │   ├── 500.html
    │   ├── apple-touch-icon-114-precomposed.png
    │   ├── apple-touch-icon-144-precomposed.png
    │   ├── apple-touch-icon-57-precomposed.png
    │   ├── apple-touch-icon-72-precomposed.png
    │   ├── apple-touch-icon-precomposed.png
    │   ├── audio # -> ../../common/audio
    │   ├── bower_components
    │   ├── common
    │   │   ├── config.js # -> ../../../common/script/config.js
    │   │   ├── directives.js # -> ../../../common/script/directives.js
    │   │   ├── habitrpg-shared.css # -> ../../../common/dist/sprites/habitrpg-shared.css
    │   │   ├── habitrpg-shared.js # -> ../../../common/dist/scripts/habitrpg-shared.js
    │   │   ├── sprites # -> ../../../common/dist/sprites
    │   │   └── userServices.js # -> ../../../common/script/userServices.js
    │   ├── community-guidelines-images
    │   ├── css
    │   ├── emails
    │   ├── favicon.ico
    │   ├── fontello
    │   ├── google280633b772b94345.html
    │   ├── google8ca65b6ff3506fb8.html
    │   ├── googlef3b1402b0e28338a.html
    │   ├── js
    │   ├── logo
    │   ├── manifest.json
    │   ├── marketing
    │   ├── page-loader.gif
    │   ├── presskit
    │   └── refresh.png
    ├── src
    └── views
```

**Comment by paglias** [comment]
@crookedneighbor the firsy options for substituting grunt-hashres that I found are https://www.npmjs.com/package/gulp-fingerprint/ https://www.npmjs.com/package/gulp-hash/ https://www.npmjs.com/package/gulp-rev-replace/

> I have some files (audio, sprites, js scripts) symlinked from common -> website/public, is that good practice? Or should it copy them on server start?

We can server them from /common, you can add anothe path to be served here https://github.com/HabitRPG/habitrpg/blob/develop/src/server.js#L118

**Comment by crookedneighbor** [comment]
**Re: calling the common files from src/server**

Awesome, that's a much better solution. 

**Re: Gulp**

Awesome, I'll try out a few and see if I can get it to work how we need it.

**Comment by taldin** [comment]
Uh oh.  Gulp is something new.   Any pointers on where to get started with it?

**Comment by taldin** [comment]
This looks like a reasonable link... http://markgoodyear.com/2014/01/getting-started-with-gulp/

**Comment by crookedneighbor** [comment]
The build process with Gulp was a little wonky, so, in the interest of time, for now I've just added in the Grunt tasks. Will make it a personal goal of mine to see if we can convert 100% to Gulp down the line.
